Possible struct def overwritten in generated C code.

**V doctor:**
```
V full version: V 0.4.6 58a8fc6.fccd7cd
OS: linux, Pop!_OS 22.04 LTS
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 2700X Eight-Core Processor

getwd: /home/txgruppi/src/REDACTED
vexe: /home/txgruppi/src/github.com/vlang/v/v
vexe mtime: 2024-06-27 15:09:16

vroot: OK, value: /home/txgruppi/src/github.com/vlang/v
VMODULES: OK, value: /home/txgruppi/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.34.1
Git vroot status: weekly.2024.26-11-gfccd7cd0
.git/config present: true

CC version: cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
thirdparty/tcc status: thirdparty-linux-amd64 40e5cbb5

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg docker-login-fix.vsh`
```v
#!/usr/bin/env -S v run

import datatypes

struct GitlabCI {
	path    string
	content string
}

fn walker(root string) chan string {
	ch_out := chan string{}

	f := fn (ch_out chan string, root string) ! {
		mut stack := datatypes.Stack[string]{}
		stack.push(root)

		ignore_list := ['.git', 'node_modules']

		for !stack.is_empty() {
			curr := stack.pop()!
			for child in ls(curr)! {
				if child in ignore_list {
					continue
				}
				childpath := join_path(curr, child)
				if is_dir(childpath) {
					stack.push(childpath)
					continue
				}
				if !child.starts_with('.gitlab-ci.') {
					continue
				}
				ch_out <- childpath
			}
		}
		ch_out.close()
	}

	spawn f(ch_out, root)

	return ch_out
}

fn loader(ch_in chan string) chan GitlabCI {
	ch_out := chan GitlabCI{}

	f := fn (ch_in chan string, ch_out chan GitlabCI) ! {
		for next := true; next; {
			next = select {
				chidlpath := <-ch_in {}
			}
		}
		ch_out.close()
	}

	spawn f(ch_in, ch_out)

	return ch_out
}

ch0 := walker(getwd())
ch1 := loader(ch0)

println(ch1)

```

**What did you expect to see?**

chan GitlabCI{cap: 0, closed: 0}

**What did you see instead?**
```
docker-login-fix.vsh:50:5: warning: unused variable: `chidlpath`
   48 |         for next := true; next; {
   49 |             next = select {
   50 |                 chidlpath := <-ch_in {}
      |                 ~~~~~~~~~
   51 |             }
   52 |         }
==================
/tmp/v_1000/docker-login-fix.01J1D3X67FYSHQAN9PYWR4T64X.tmp.c:24985: warning: assignment from incompatible pointer type
/tmp/v_1000/docker-login-fix.01J1D3X67FYSHQAN9PYWR4T64X.tmp.c:24987: error: cannot convert 'struct sync__Channel *' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**Maybe helpful information**

It looks like the issue is because of the `f` function inside `walker` and `loader` and the generated C code has the `thread_arg_f` definition overwritten because of the other function with the same name.

If I change the name of one of the `f` functions to `ff`, for example, the code works fine.
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/spawn_and_go.v
+++ b/vlib/v/gen/c/spawn_and_go.v
@@ -26,9 +26,14 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 	mut handle := ''
 	tmp := g.new_tmp_var()
 	mut expr := node.call_expr
-	mut name := '${expr.name}_${expr.pos.pos}'
+	mut name := expr.name
 	mut use_tmp_fn_var := false
 	tmp_fn := g.new_tmp_var()
+
+	if expr.is_fn_var {
+		// generate a name different for same var fn name declared in another scope
+		name = '${name}_${node.pos.pos}'
+	}
 
 	if expr.concrete_types.len > 0 {
 		name = g.generic_fn_name(expr.concrete_types, name)

--- a/vlib/v/gen/c/spawn_and_go.v
+++ b/vlib/v/gen/c/spawn_and_go.v
@@ -26,7 +26,7 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 	mut handle := ''
 	tmp := g.new_tmp_var()
 	mut expr := node.call_expr
-	mut name := expr.name
+	mut name := '${expr.name}_${expr.pos.pos}'
 	mut use_tmp_fn_var := false
 	tmp_fn := g.new_tmp_var()
 

--- a/vlib/v/tests/chan_same_fn_name_test.v
+++ b/vlib/v/tests/chan_same_fn_name_test.v
@@ -1,0 +1,24 @@
+fn a() chan string {
+	ch_out := chan string{}
+	f := fn (a chan string) {
+		a <- 'foo'
+	}
+	spawn f(ch_out)
+	return ch_out
+}
+
+fn b(ch_in chan string) string {
+	f := fn (a chan string, b chan string) {
+		val := <-a
+		{}
+		b <- val
+	}
+	ch_out := chan string{}
+	spawn f(ch_in, ch_out)
+	return <-ch_out
+}
+
+fn test_main() {
+	ch0 := a()
+	assert b(ch0) == 'foo'
+}

